presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateFactory.java

```diff
@@ -0,0 +1,186 @@
+    {
+        return new SingleLongMinMaxState(type, methodHandle);
+    }
+
+    @Override
+    public Class<? extends LongMinMaxState> getSingleStateClass()
+    {
+        return SingleLongMinMaxState.class;
+    }
+
+    @Override
+    public LongMinMaxState createGroupedState()
+    {
+        return new GroupLongMinMaxState(type, methodHandle);
+    }
+
+    @Override
+    public Class<? extends LongMinMaxState> getGroupedStateClass()
+    {
+        return GroupLongMinMaxState.class;
+    }
+
+    public static class GroupLongMinMaxState
+            extends AbstractGroupedAccumulatorState
+            implements LongMinMaxState
+    {
+        private final Type type;
+        private final MethodHandle methodHandle;
+        private final LongBigArray values = new LongBigArray();
+        private final BooleanBigArray nulls = new BooleanBigArray(true);
+
+        public GroupLongMinMaxState(Type type, MethodHandle methodHandle)
+        {
+            this.type = type;
+            this.methodHandle = methodHandle;
+        }
+
+        @Override
+        public void ensureCapacity(long size)
+        {
+            values.ensureCapacity(size);
+            nulls.ensureCapacity(size);
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return values.sizeOf() + nulls.sizeOf();
+        }
+
+        @Override
+        public Type getType()
+        {
+            return type;
+        }
+
+        @Override
+        public MethodHandle getMethodHandle()
+        {
+            return methodHandle;
+        }
+
+        @Override
+        public long getLong()
+        {
+            return values.get(getGroupId());
+        }
+
+        @Override
+        public void setLong(long value)
+        {
+            values.set(getGroupId(), value);
+        }
+
+        @Override
+        public boolean isNull()
+        {
+            return nulls.get(getGroupId());
+        }
+
+        @Override
+        public void setNull(boolean value)
+        {
+            nulls.set(getGroupId(), value);
+        }
+    }
+
+    public static class SingleLongMinMaxState
+            implements LongMinMaxState
+    {
+        private final Type type;
+        private final MethodHandle methodHandle;
+        private long value;
+        private boolean isNull;
+        private static final int estimatedSize = ClassLayout.parseClass(SingleLongMinMaxState.class).instanceSize();
+
+        public SingleLongMinMaxState(Type type, MethodHandle methodHandle)
+        {
+            this.type = type;
+            this.methodHandle = methodHandle;
+            this.value = 0;
+            this.isNull = true;
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return estimatedSize;
+        }
+
+        @Override
+        public Type getType()
+        {
+            return type;
+        }
+
+        @Override
+        public MethodHandle getMethodHandle()
```

Comment:
My comment about using LongState instead of LongMaxMinState seems to have been lost, and I can't find Chris' reply now, but I wanted to point out I think the method handle shouldn't be stored inside the state object. It should be bound into the method handle for the input function at specialization time, instead. 

The type is something that could be added to LongState, etc., and seems like a generally useful thing to do anyway. Once you do that, you can get rid of these extra state classes, factories, etc.